src/butil/single_iobuf.cpp

```diff
@@ -0,0 +1,279 @@
+        LOG(ERROR) << "SingleIOBuf reallocate_downward failed. Block cannot be null!";
+        errno = EINVAL;
+        return NULL;
+    }
+    char* old_p = ref.block->data + ref.offset;
+    uint32_t old_size = ref.length;
+    IOBuf::Block* b = alloc_block_by_size(new_size);
+    if (!b) {
+        return NULL;
+    }
+    char* new_p = b->data + b->size;
+    memcpy_downward(old_p, old_size,
+                    new_p, new_size,
+                    in_use_back, in_use_front);
+    ref.block->dec_ref();
+    _cur_ref.offset = b->size;
+    _cur_ref.length = new_size;
+    _cur_ref.block = b;
+    b->size += new_size;
+    b->inc_ref();
+    return new_p;
+}
+
+const void* SingleIOBuf::get_begin() const {
+    if (_cur_ref.block) {
+        return _cur_ref.block->data + _cur_ref.offset;
+    }
+    return NULL;
+}
+
+uint32_t SingleIOBuf::get_length() const {
+    return _cur_ref.length;
+}
+
+void SingleIOBuf::reset() {
+    if (_cur_block) {
+        if (_block_size == 0) {
+            iobuf::release_tls_block(_cur_block);
+        } else {
+            _cur_block->dec_ref();
+            _block_size = 0;
+        }
+        _cur_block = NULL;
+    }
+    if (_cur_ref.block != NULL) {
+        _cur_ref.block->dec_ref();
+    }
+    _cur_ref.offset = 0;
+    _cur_ref.length = 0;
+    _cur_ref.block = NULL;
+}
+
+bool SingleIOBuf::assign(const IOBuf& buf, uint32_t msg_size) {
+    if (BAIDU_UNLIKELY(buf.length() < msg_size)) {
+        LOG(ERROR) << "Fail to dump_mult_iobuf msg_size:" << msg_size
+            << " from source iobuf of size:" << buf.length();
+        return false;
+    }
+    reset();
+    if (BAIDU_UNLIKELY(msg_size == 0)) {
+        return true;
+    }
+    const IOBuf::BlockRef& ref = buf._front_ref();
+    if (ref.length >= msg_size) {
+        _cur_ref.offset = ref.offset;
+        _cur_ref.length = msg_size;
+        _cur_ref.block = ref.block;
+        _cur_ref.block->inc_ref();
+        return true;
+    } else {
+        IOBuf::Block* b = alloc_block_by_size(msg_size);
+        if (!b) {
+            return false;
+        }
+        char* out = b->data + b->size;
+        const size_t nref = buf.backing_block_num();
+        uint32_t last_len = msg_size;
+        for (size_t i = 0; i < nref && last_len > 0; ++i) {
+            const IOBuf::BlockRef& r = buf._ref_at(i);
+            uint32_t n = std::min(r.length, last_len);
+            iobuf::cp(out, r.block->data + r.offset, n);
+            last_len -= n;
+            out += n;
+        }
+        _cur_ref.offset = b->size;
+        _cur_ref.length = msg_size;
+        _cur_ref.block = b;
+        b->size += msg_size;
+        _cur_ref.block->inc_ref();
+        return true;
+    }
+}
+
+void SingleIOBuf::append_to(IOBuf* buf) const {
+    if (buf && _cur_ref.block) {
+        buf->_push_back_ref(_cur_ref);
+    }
+}
+
+int SingleIOBuf::assign_user_data(void* data, size_t size, void (*deleter)(void*)) {
+    reset();
```

Comment:
This reset may have side effects. Should it be executed after the condition check?